### DISCUSSION TITLE
confused why rez-pip creates python variant

### DISCUSSION THREAD

**Discussion by keith-lackey-steamroller-tech** [AcademySoftwareFoundation/rez#1425] [Q&A]
When doing a rez-pip install it seems to make a variant that includes the current python version. 

For example:

`rez-pip --install pymongo`

I end up with something like this in my package.py

`
variants = [
    ['platform-windows', 'arch-64bit', 'python-3.11.1']
]
`

Python 3.11.1 just happens to be the version of Python that was running whenever it was installed but it is not what version the package is compatible with. 

If you look look at their PyPi page it states that it requires python-3.7+ which would suggest that it would be compatible with any version going back to Python 3.7

[https://pypi.org/project/pymongo/](url)

The funny thing is that the requires is missing Python completely.

`requires = ['dnspython-1.16.0+<3.0.0']`

I would assume that the Python version would not be included in the variant and that it would be part of the requirements instead.

```
requires = ['dnspython-1.16.0+<3.0.0', 'python-3.7+']

variants = [
    ['platform-windows', 'arch-64bit']
]
```

Is this assumption correct?

Should I be using the results of the `rez-pip` as a jumping off point for the package and reorganizing it once its been created?

Is there a better way to control this?

Or am I missing altogether?

Thanks in advance!

**Comment by instinct-vfx** [comment]
pymongo indeed states 3.7+ as a requirement, that does include 3.11 though and 3.11 is indeed that newest supported version.

![image](https://user-images.githubusercontent.com/243828/210168058-514b3187-8220-493a-b62f-7c00f6ff79f7.png)

As pymongo contains binary components they need to provide different wheels for different python versions as they are not ABI compatible.

These are also added to the variants and not requires to allow having the same version of pymongo for different python versions as variants.

If you want to add a version for a specific python version you need to tell rez-pip by supplying `--python-version 3.7` for example.

**Reply by keith-lackey-steamroller-tech** [reply]
Thanks for your prompt answer and I think I understand what your saying.

So is the standard practice to iterate over all python packages and pip install them when looking to support a different version of python?

On a side note, I have run into a problem though when trying to use the `--python-version` flag because I have pip installed as a separate rez package from the python rez package and think I found a bug in the pip.py @ line 212 inside of the rez library

`py_exe, "-E", "-s", "-c", "import pip, sys; sys.stdout.write(pip.__version__)"],`

The `-E` tells python to ignore `PYTHON*` environmental variables such as PYTHONPATH when starting up. Since I have pip installed as a separate rez package and in that package it uses the PYTHONPATH to tell python where pip lives, it can't find it. I think the whole point of that section of the rez code is to support pip installed as a package. If I remove that `-E`, the `--python-version` flag works as expected. I'm not entirely sure why those flags were put in there but it looks like it might have had unintended consequences.

**Reply by instinct-vfx** [reply]
For the line you are referring to check the above docstring. This is done intentional to make sure we are isolating as much as possible from anything outside rez and also ensure we are not ending up with pip or python from somewhere else. This line specifically runs python to get the version of pip that is included with this specific installation and not other packages or installations.

Here is more information on `rez-pip` and also specifically on how it detects and chooses pip versions: https://github.com/AcademySoftwareFoundation/rez/wiki/Pip#which-pip-will-be-used?  Note the order of selecting pip from different sources.
